### PR TITLE
refactor shard optimizer to support unification of dynamic and static…

### PR DESCRIPTION
**Pull request by winter-wang**
### PR Category

Auto Parallel

### PR Types

New features

### Description

- support shard optimizer for pir mode.
    - the optimizer op in stage1:
    ```python
        (%0) = "pd_op.data" () {dtype:(pd_op.DataType)float32,name:"embedding_0.w_0.dist_beta2_pow_acc_0",...} : () -> pd_dist.tensor<1xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1]>
        (%1) = "pd_op.data"() {dtype:(pd_op.DataType)float32,name:"embedding_0.w_0.dist_beta1_pow_acc_0",...} : () -> pd_dist.tensor<1xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1]>
        (%2) = "pd_op.data" () {dtype:(pd_op.DataType)float32,name:"embedding_0.w_0.dist_moment2_0",...} : () -> pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>
        (%3) = "pd_op.data" [id:963] () {dtype:(pd_op.DataType)float32,name:"embedding_0.w_0.dist_moment1_0",...} : () -> pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>
        (%1147, %1148, %1149, %1150, %1151, %1152) = "pd_op.adamw_"(%171, %906, %907, %3, %2, %1, %0, <<NULL VALUE>>, <<NULL VALUE>>, %1144, %1145, %1146) {...omit attribute} :
             ( # operand type 
                  # param
                  pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>, 
                  # grad
                  pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>, 
                  # learning rate
                  pd_dist.tensor<f32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[]>,
                  # moment1
                  pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>,
                   # moment2
                  pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>, 
                  # beta1_pow
                  pd_dist.tensor<1xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1]>, 
                  #beta2_pow
                  pd_dist.tensor<1xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1]>,
                   #master_param
                  <<NULL TYPE>>,
                  # skip_update
                  <<NULL TYPE>>,
                  # beta1
                  pd_dist.tensor<1xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1]>,
                  # beta2
                  pd_dist.tensor<1xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1]>,
                  # epsilon
                  pd_dist.tensor<1xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1]>) -> 
                # ouptuts type
                  pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>, 
                  pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>, 
                  pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>, 
                  pd_dist.tensor<1xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1]>,
                  pd_dist.tensor<1xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1]>, <<NULL TYPE>>
        # reshard parameter to origin dist attr
       (%1153) = "pd_op.c_allgather"  (%1147) {} : 
                (pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[0,1]>) -> 
                 pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1,1]>
        # update global parameter
        "pd_op.assign_out_" (%1153, %168) {} :
                 (pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1,1]>, 
                  pd_dist.tensor<32000x1024xf32, mesh_shape:[2,2],process_ids:[0,1,2,3],dims_mappings:[-1,1]>)
    ```
- support shard startup program.
- support p_to_s reshard function.
- fix the bug in r_to_s reshard function. 
- add monkey_patch_value_in_dist called in distribute init function. 

### Other
pcard-67164

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/api.py [hunk 1]
```diff
@@ -976,58 +973,33 @@ def _shard_accumulator(self, param):
...
-                    param, param.process_mesh, new_placement
-                )
+        if param.is_dist() and isinstance(
+            self._shard_fn, (ShardingStage1, ShardingStage2)
+        ):
+            new_placement = param.placements
+            new_placement[self._sharding_mesh_axis] = dist.Replicate()
+            out_param = dist.reshard(param, param.process_mesh, new_placement)
+            if in_pir_mode():
+                paddle.assign(out_param, param)
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/api.py]: 这样 assign 有 D2D 拷贝，而且会开辟一个新显存空间，让静半sharidng 训练时，同一个参数在内存上同时存在：old_replicated_param, new_shard_param, new_replicated_param 3 个副本，增大显存峰值，无法达到sharding opt 降低显存的目标。

需要通过一个方式  share_data_with(new_replicated_param,  old_replicated_param)

Comment (winter-wang) [reply python/paddle/distributed/auto_parallel/api.py]: done, thanks!

File: python/paddle/distributed/auto_parallel/api.py [hunk 2]
```diff
@@ -826,9 +827,23 @@ def replicate_layer_params_and_buffers(
...
 
 
+def get_value_placement(value):
+    dist_attr = value.dist_attr()
+    assert dist_attr is not None, "Can't get placement for a dense value."
+    mesh = dist_attr.process_mesh
+    dims_mapping = dist_attr.dims_mapping
+    partial_status = dist_attr.partial_status
+    palcements = to_placements(dims_mapping, mesh, partial_status)
+    return palcements
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py]: We can add `placement` as  Value's property in the future to make it dynamic-static consistent

Comment (winter-wang) [reply python/paddle/distributed/auto_parallel/api.py]: done, thanks!

File: python/paddle/distributed/auto_parallel/api.py [hunk 3]
```diff
@@ -970,64 +982,42 @@ def _shard_accumulator(self, param):
...
+                )
 
     def _reset_placements(self, param):
-        if param.is_dist():
-            if isinstance(self._shard_fn, (ShardingStage1, ShardingStage2)):
+        if param.is_dist() and isinstance(
+            self._shard_fn, (ShardingStage1, ShardingStage2)
+        ):
+            # in pir mode, reshard pass will automatically handle inplace case, so no extra work is required here.
+            if not isinstance(param, pir.Value):
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/api.py:983]: pir 自动 inplace，但 reshard 还是要手动调的吧

Comment (winter-wang) [reply python/paddle/distributed/auto_parallel/api.py:983]: reshard已经在reshard_pass中，全部调到了最前面

**Review by JZ-LIANG — APPROVED** [review]
LGTM

**Review by sneaxiy — APPROVED** [review]
LGTM for `python/paddle/distributed/__init__.py`

**Pull request merged**